cpp/src/arrow/flight/sql/odbc/tests/type_info_test.cc

```diff
@@ -0,0 +1,1262 @@
+                      SQL_FALSE,              // expectedCaseSensitive
+                      SQL_SEARCHABLE,         // expectedSearchable
+                      SQL_FALSE,              // expectedUnsignedAttr
+                      SQL_FALSE,              // expectedFixedPrecScale
+                      NULL,                   // expectedAutoUniqueValue
+                      std::wstring(L"time"),  // expectedLocalTypeName
+                      NULL,                   // expectedMinScale
+                      NULL,                   // expectedMaxScale
+                      SQL_DATETIME,           // expectedSqlDataType
+                      NULL,   // expectedSqlDatetimeSub, driver returns NULL for Ver2
+                      NULL,   // expectedNumPrecRadix
+                      NULL);  // expectedIntervalPrec
+
+  // No more data
+  ret = SQLFetch(this->stmt);
+  EXPECT_EQ(ret, SQL_NO_DATA);
+
+  this->disconnect();
+}
+
+TEST_F(FlightSQLODBCMockTestBase, TestSQLGetTypeInfoTimestamp) {
+  this->connect();
+
+  SQLRETURN ret = SQLGetTypeInfo(this->stmt, SQL_TYPE_TIMESTAMP);
+  EXPECT_EQ(ret, SQL_SUCCESS);
+
+  // Check timestamp data type
+  ret = SQLFetch(this->stmt);
+  EXPECT_EQ(ret, SQL_SUCCESS);
+
+  checkSQLGetTypeInfo(this->stmt,
+                      std::wstring(L"timestamp"),  // expectedTypeName
+                      SQL_TYPE_TIMESTAMP,          // expectedDataType
+                      32,                          // expectedColumnSize
+                      std::wstring(L"'"),          // expectedLiteralPrefix
+                      std::wstring(L"'"),          // expectedLiteralSuffix
+                      std::nullopt,                // expectedCreateParams
+                      SQL_NULLABLE,                // expectedNullable
+                      SQL_FALSE,                   // expectedCaseSensitive
+                      SQL_SEARCHABLE,              // expectedSearchable
+                      SQL_FALSE,                   // expectedUnsignedAttr
+                      SQL_FALSE,                   // expectedFixedPrecScale
+                      NULL,                        // expectedAutoUniqueValue
+                      std::wstring(L"timestamp"),  // expectedLocalTypeName
+                      NULL,                        // expectedMinScale
+                      NULL,                        // expectedMaxScale
+                      SQL_DATETIME,                // expectedSqlDataType
+                      SQL_CODE_TIMESTAMP,          // expectedSqlDatetimeSub
+                      NULL,                        // expectedNumPrecRadix
+                      NULL);                       // expectedIntervalPrec
+
+  // No more data
+  ret = SQLFetch(this->stmt);
+  EXPECT_EQ(ret, SQL_NO_DATA);
+
+  this->disconnect();
+}
+
+TEST_F(FlightSQLODBCMockTestBase, TestSQLGetTypeInfoTimestampVer2) {
+  this->connect(SQL_OV_ODBC2);
+
+  SQLRETURN ret = SQLGetTypeInfo(this->stmt, SQL_TIMESTAMP);
```

Comment:
We should test SQL_TYPE_* in ODBC 2.x as well (and test SQL_* datetime in ODBC 3.x).
Also test SQL_ALL_TYPES in ODBC 2.x

Reply:
Good catch, I have added tests for these 3 types